pkg/thing/interactors/register_thing.go

```diff
@@ -82,25 +82,27 @@ func (i *ThingInteractor) Register(authorization, id, name string) error {
 	if err != nil {
 		errReply := i.reply(id, "", err)
 		if errReply != nil {
-			i.logger.Error(errReply)
+			i.logger.Debug(err)
+			return errReply
 		}
-
 		return err
 	}
 
 	// Get the id generated as a token and send in the response
 	token, err := i.thingProxy.Create(id, name, authorization)
-	errReply := i.reply(id, token, err)
-	if errReply != nil {
-		i.logger.Error(errReply)
-		return errReply
+	if err != nil {
+		errReply := i.reply(id, token, err)
+		if errReply != nil {
+			i.logger.Debug(err)
+			return errReply
+		}
+		return err
 	}
 
-	err = i.connectorPublisher.SendRegisterDevice(id, name)
+	err = i.reply(id, token, err)
 	if err != nil {
 		i.logger.Error(err)
-		return err
 	}
 
-	return nil
+	return i.connectorPublisher.SendRegisterDevice(id, name)
 }
```

Comment:
We should avoid returning statements such as this one. Modifying [Dave Chaney's example](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) to our context :point_down: 

> If `i.connectorPublisher.SendRegisterDevice(id, name)` returns an error, then `Register` will return the error to its caller, who will probably do the same, and so on. At the top of the program the main body of the program will print the error to the screen or a log file, and all that will be printed is: "No such file or directory," for example. In other words, we cannot tell where the original error came from.

Here's an example :smile: 

```suggestion
	err = i.connectorPublisher.SendRegisterDevice(id, name)
	if err != nil {
		i.logger.Error(err)
		return err
	}

	return nil
}
```

However, if we must return `i.connectorPublisher.SendRegisterDevice(id, name)`, it'd be nice to formalize and standardize when and why to do this (such as in [here](https://github.com/CESARBR/knot-babeltower/blob/master/pkg/thing/delivery/http/thing_proxy.go#L201)).

Reply:
+1
You're right. However, in this case, we will get the same result.

The point about handling errors as Dave Cheney suggested is that we have to append contextual information to the error before return it (@esdrasjnr I strongly suggest you read the entire session of the article @lcbm mentioned).

In a lot of cases, we don't append any contextual information to the errors. Let's change this incrementally.

Reply:
@lcbm @joaoaneto Excellent article, thanks! In order to add some context to errors, I made some changes in code. Could you please check if it makes sense? :smiley: